src/heartbeat.c

```
@@ -0,0 +1,188 @@
+/*
+ * Copyright 2019 Xaptum, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "heartbeat.h"
+
+#include <time.h>
+#include <uv.h>
+
+#include "icmp.h"
+#include "ip.h"
+#include "log.h"
+
+static void
+on_request_timer(uv_timer_t* timer);
+
+static void
+on_reply_timer(uv_timer_t* timer);
+
+static void
+on_write(struct enftun_crb* crb)
+{
+    if (crb->status)
+        enftun_log_error("PING: failed to send heartbeat reply: %d\n",
+                         crb->status);
```

Rather than ignoring the error, we should have enftun reconnect.